Use official CASDK client to connect WebAPI

YaSuenag opened the pull request:
Carbon Aware SDK has published official WebAPI client libraries includes NPM. So I'd like to propose to use it to communicate with WebAPI.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Enhancement (project structure, spelling, grammar, formatting)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.


### A description of the changes proposed in the Pull Request

This PR adds dependency to official NPM library for WebAPI 1.0.0: https://github.com/Green-Software-Foundation/carbon-aware-sdk/pkgs/npm/casdk-client/320495557

YaSuenag committed:
Use official CASDK client to connect WebAPI

Add dependency to official NPM library for WebAPI 1.0.0:
https://github.com/Green-Software-Foundation/carbon-aware-sdk/pkgs/npm/casdk-client/320495557

danuw commented:
can we update the workflow to include manual trigger for now of the workflow so we an ensure tests pass?

YaSuenag commented:
@danuw CC: @nttDamien
I tweaked my PR - CI workflow works fine: https://github.com/YaSuenag/if-casdk-plugin/actions/runs/12961068546

I believe all tests would be passed if you approve the workflow on this PR.

YaSuenag commented:
@danuw 
CI workflow was failed due to incorrect branch reference, so I fixed.
Can you approve workflow run again?

YaSuenag commented:
@danuw 
Can you wait review? Workflow is done, but global configuration does not seem to be applied. I will investigate.

YaSuenag commented:
I tested current HEAD of `main`, it has same problem that IF cannot read `global-config` parameters. I try to fix it in this PR.

```
$ npm run if-run -- --manifest ../casdk.yaml

> @grnsft/if@1.0.0 if-run
> npx ts-node src/if-run/index.ts --manifest ../casdk.yaml

[2025-01-31 09:06:37.913 AM] info:
Graduated Project

This project is a Graduated Project, supported by the Green Software Foundation. The publicly available version documented in the README is trusted by the GSF. New versions of the project may be released, or it may move to the Maintained or Retired Stage at any moment.

[2025-01-31 09:06:39.506 AM] warn:
  You are using plugin if-casdk-plugin which is not part of the Impact Framework standard library. You should do your own research to ensure the plugins are up to date and accurate. They may not be actively maintained.

[2025-01-31 09:06:39.538 AM] TypeError:         Cannot read properties of undefined (reading 'regions')
    at getCarbonRating (/home/yasuenag/github/if-casdk-plugin/src/lib/casdk/index.ts:68:34)
    at /home/yasuenag/github/if-casdk-plugin/src/lib/casdk/index.ts:43:32
    at Array.map (<anonymous>)
    at Object.execute (/home/yasuenag/github/if-casdk-plugin/src/lib/casdk/index.ts:41:14)
    at computeNode (/home/yasuenag/test/if/if/src/if-run/lib/compute.ts:193:36)
    at traverse (/home/yasuenag/test/if/if/src/if-run/lib/compute.ts:35:11)
    at computeNode (/home/yasuenag/test/if/if/src/if-run/lib/compute.ts:98:12)
    at compute (/home/yasuenag/test/if/if/src/if-run/lib/compute.ts:221:9)
    at impactEngine (/home/yasuenag/test/if/if/src/if-run/index.ts:47:39)
    at processTicksAndRejections (node:internal/process/task_queues:105:5)

# start
name: casdk
description: if-casdk-plugin test
initialize:
  plugins:
    casdk:
      path: if-casdk-plugin
      method: CaSdkPlugin
      global-config:
        regions: uksouth,westeurope
        baseUrl: https://carbon-aware-api.azurewebsites.net
execution:
  status: fail
  command: >-
    /home/yasuenag/.npm/_npx/1bf7c3c15bf47d04/node_modules/.bin/ts-node
    /home/yasuenag/test/if/if/src/if-run/index.ts --manifest ../casdk.yaml
  environment:
    if-version: 1.0.0
    os: linux
    os-version: 6.11.11-300.fc41.x86_64
    node-version: 22.11.0
    date-time: 2025-01-31T00:06:39.501Z (UTC)
    dependencies:
      - '@babel/core@7.22.10'
      - '@babel/preset-typescript@7.23.3'
      - '@commitlint/cli@18.6.0'
      - '@commitlint/config-conventional@18.6.0'
      - '@grnsft/if-core@0.0.28'
      - '@jest/globals@29.7.0'
      - '@types/jest@29.5.8'
      - '@types/js-yaml@4.0.9'
      - '@types/luxon@3.4.2'
      - '@types/node@20.9.0'
      - axios-mock-adapter@1.22.0
      - axios@1.7.7
      - cross-env@7.0.3
      - csv-parse@5.5.6
      - csv-stringify@6.4.6
      - fixpack@4.0.0
      - gts@5.2.0
      - husky@8.0.3
      - >-
        if-casdk-plugin@v0.0.2 extraneous ->
        file:../../../../github/if-casdk-plugin
      - jest@29.7.0
      - js-yaml@4.1.0
      - lint-staged@15.2.10
      - luxon@3.4.4
      - release-it@16.3.0
      - rimraf@5.0.5
      - ts-command-line-args@2.5.1
      - ts-jest@29.1.1
      - typescript-cubic-spline@1.0.1
      - typescript@5.2.2
      - winston@3.11.0
      - zod@3.23.8
  error: 'TypeError: Cannot read properties of undefined (reading ''regions'')'
tree:
  children:
    child-0:
      defaults:
        cpu/thermal-design-power: 100
      pipeline:
        compute:
          - casdk
      inputs:
        - timestamp: 2023-07-06T00:00
          duration: 1
          cpu/utilization: 20.5
        - timestamp: 2023-07-06T00:01
          duration: 1
          cpu/utilization: 80
        - timestamp: 2023-07-06T00:02
          duration: 1
          cpu/utilization: 20

# end
```

YaSuenag commented:
@danuw CC: @nttDamien
I fixed all of problems in if-casdk-plugin and it works fine on my fork. I think it should work on GHA on this repo, but it might be tweaked a bit - anyway I want to see the result when @danuw approve the workflow in this PR.
https://github.com/YaSuenag/if-casdk-plugin/actions/runs/13067352302

Original manifest file generated in the workflow is no longer correct. Plugin configuration should be described in `config` section rather than `global-config`. And also `regions` are expected string array rather than single string delimiterd by `,`.

Test result with IF was not checked, so this PR checks and exit with `1` to notify error to GHA runner when the status is not `success`.

This plugin would output timestamp of emission data, but it overrides timestamp of input data, so this PR renamed it to `casdk-timestamp`.

I updated source code structure as following:

```
src/
├── __tests__
│   └── unit
│       └── lib
│           └── best-emissions-data-for-locations-by-time.ts
├── index.ts
└── lib
    └── best-emissions-data-for-locations-by-time
        └── index.ts
```

`src/index.ts` has entry point of the endpoint as following:

```typescript
export {BestEmissionsDataForLocationsByTime} from './lib/best-emissions-data-for-locations-by-time';
```

So you can add new plugin function under `src/lib`, and it should be added to `src/index.ts` to expose when you want to new WebAPI wrapper to this plugin.

danuw commented:
Thank you very much @YaSuenag

danuw commented:
Next we will probably need to update the readme and what I mentioned about IF core dependencies that need to be adjusted/refreshed and fixed as they are now available in IF Core.

Thanks again and catch up soon
